ddtrace/llmobs/_llmobs.py

```diff
@@ -273,9 +292,39 @@ def annotation_context(
                         This argument is only applicable to LLM spans.
         :param name: Set to override the span name for any spans annotated within the returned context.
         """
-        return AnnotationContext(
-            cls._instance.tracer, lambda span: cls.annotate(span, tags=tags, prompt=prompt, _name=name)
-        )
+        # id to track an annotation for registering / de-registering
+        annotation_id = rand64bits()
+
+        def get_annotations_context_id():
+            current_ctx = cls._instance.tracer.current_trace_context()
+            # default the context id to the annotation id
+            ctx_id = annotation_id
+            if current_ctx is None:
+                current_ctx = Context(is_remote=False)
+                current_ctx._set_baggage_item(ANNOTATIONS_CONTEXT_ID, ctx_id)
+                cls._instance.tracer.context_provider.activate(current_ctx)
+            elif not current_ctx._get_baggage_item(ANNOTATIONS_CONTEXT_ID):
+                current_ctx._set_baggage_item(ANNOTATIONS_CONTEXT_ID, ctx_id)
+            else:
+                ctx_id = current_ctx._get_baggage_item(ANNOTATIONS_CONTEXT_ID)
+            return ctx_id
+
+        def register_annotation():
+            with cls._instance._annotation_context_lock:
+                ctx_id = get_annotations_context_id()
+                cls._instance._annotations.append(
+                    (annotation_id, ctx_id, {"tags": tags, "prompt": prompt, "_name": name})
+                )
+
+        def deregister_annotation():
+            with cls._instance._annotation_context_lock:
+                for i, (key, _, _) in enumerate(cls._instance._annotations):
+                    if key == annotation_id:
+                        cls._instance._annotations.pop(i)
+                        return
+                    log.debug("Failed to pop annotation context")
```

Comment:
this should be in the `else` of the `for` loop right (if it falls through)? This will log whenever a key doesn't match which is expected when having two threads of execution

Reply:
Ah yes, thanks for catching